### ISSUE TITLE
Allow StringLists with suggestions instead of hard restrictions

### ISSUE THREAD

**Issue by mknierman** [OpenMS/OpenMS#1949]
I needed to add mods to my search and I updated the unimod.xml file with these mods but I can not see them listed in the xtandem adapter in KNIME.  If I use the xtandem adapter in toppas and modify the unimod.xml I see the changes from the unimod.xml file.  Is there an issue with building the mod list in the KNIME xtandem adapter?
Thanks,

BTW great tools!

Mike

**Comment by timosachsenberg** [comment]
Hi @mknierman ,
currently, it is (unfortunately) not possible to update the KNIME nodes based on changes in the unimod XML. This is mostly by design as KNIME tool dialogs in KNIME are generated once and remain static. In the use case of editing the unimod.xml, the tool automatically retrieves the updated modifications but the KNIME node used to configure it underlying OpenMS tool remains the same. We would need to chance the restrictions (i.e., the allowed options of a tool parameter) in the KNIME node, too. @jpfeuffer do you know if there might be a possibility in a newer version of KNIME to e.g. adapt the tool "restrictions" based on external files (e.g. hook somehow into the node configure dialog)?

**Comment by lars20070** [comment]
The inability to use a locally modified `unimod.xml` file would prevent us to switch from `TOPPAS` to Knime. Very important issue for us too.

FYI, @oliverschilling.

**Comment by jpfeuffer** [comment]
To be clear: You basically want the configuration dialog to change based on an external file that you give somewhere?

**Comment by jpfeuffer** [comment]
I will ask the KNIME guys if this is possible. An easy workaround would be to read the file in another node, select the modifications there and then pass the selected values as a workflow variable.

**Comment by hendrikweisser** [comment]
> An easy workaround would be to read the file in another node, select the modifications there (from a nice dropdown menu, e.g. "Value Selection Quickform" node ) and then pass the selected values as a workflow variable, which you could btw reuse in every searchengine then.

I don't think this would currently work with OpenMS, because the parameter values are validated by the TOPP tool. Wouldn't you get an error about an "illegal parameter value" or something to that extent?

**Comment by timosachsenberg** [comment]
@jpfeuffer yes, basically the list of modifications that are available in the String List Editor would need to be extended

**Comment by mknierman** [comment]
Yes that would be correct.  There is a unimod.xml file that is installed with the openms node.  Making changes to this file should update the list of modifications presented to the user.   It works this way in TOPPAS.  

Mike

> On Apr 19, 2016, at 5:28 AM, jpfeuffer <notifications@github.com> wrote:
> 
> To be clear: You basically want the configuration dialog to change based on an external file that you give somewhere?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/OpenMS/OpenMS/issues/1949#issuecomment-211823262>

Mike Knierman
mknierman@comcast.net

**Comment by jpfeuffer** [comment]
Okay I thought a bit about what we would need for this:
- in OpenMS: a custom cmake script that marks all the options in the CTDs that were read from a file and instead of listing all the elements for the dropdown menu, just somehow describe how to parse that file to get that list
- in GKN: 
   - a subclass of the NodeConfiguration interface that can do different parsing tasks (or calls an OpenMS tool that does the parsing) at runtime [possible, talked to KNIME guys]
   - during the build of the nodes we need to recognize when such an option is present and use the corresponding NodeConfiguration subclass

The problem is, as @hendrikweisser said, that the CLI/config parsing/validating is not done in the same tool that runs the algorithm and has access to the ModificationDB anymore.

However we could emulate that behaviour by having a preceding ListModifications node that returns the list of possible modifications (from the unimod.xml and PSI.obo in the share; that is btw another problem: the modifications are actually parsed from two files). Then the user selects them before the actual SearchEngine configuration dialogue and passes them as variables.
For the xml it is easily possible with a series of nodes, see KNIME workflow attachment (no guarantees):
[unimod_reading.zip](https://github.com/OpenMS/OpenMS/files/226518/unimod_reading.zip)

**Comment by timosachsenberg** [comment]
Hmm this seems possible but also pretty complex on the first sight.
Let me brainstorm a bit -  I don't know if this is possible:
- we add a tag to the restrictions in the ctd that specifies that the restrictions are e.g. "dynamic" which means that additional entries can be provided by the user.
-In the StringListDialog in KNIME we would still have all entries from the restrictions but could provide the additional possibility to add a user defined string.
As the user added the modification himself to the unimod.xml I see no problem in letting him enter it manually.
This would be imho less complex and - as it is only used in the search engine adapters - probably an easy solution. Big disadvantage for this - and all other proposed methods is that we can't automatically bundle and share the workflow anymore.

**Comment by jpfeuffer** [comment]
Could we then just have an additional parameter in the search engines for user-defined mods, where the user has to make sure that they are present in the files that are scanned for the ModDB. And no restrictions are put on there?

**Comment by timosachsenberg** [comment]
hmm I would not do this on the tool level as it complicates the tool interface to solve a KNIME / GKN issue.

If we want to solve the workflow sharing issue we would need:
- To allow specifying external unimox /psmod files in search engine adapters which default to the ones in share.
- A mechanism to propagate the entries to the KNIME node configuration dialog

**Comment by timosachsenberg** [comment]
Another option is to just add e.g. 10 usermod1-10 to unimod.xml entries before we generate the nodes. Easiest solution but doesn't solve data sharing issue.

**Comment by jpfeuffer** [comment]
Interesting.. that's actually quite nice. About data sharing.. is this possible with TOPPAS? You would also have to ship the modified unimod.xml right?

**Comment by timosachsenberg** [comment]
yes same problem there but it is more pressing in KNIME if we want to share workflows

**Comment by jpfeuffer** [comment]
Yes. I think a param for a user-defined path that defaults to the share should solve it right?
Problem is that it is not just one file.

**Comment by timosachsenberg** [comment]
yes if combined with the additional usermod1-n entries in the unimod it would work

**Comment by timosachsenberg** [comment]
should we then use the PSI-mod file or the unimod file?

**Comment by hendrikweisser** [comment]
> Another option is to just add e.g. 10 usermod1-10 to unimod.xml entries before we generate the nodes. 

I'm not a fan of this, because you might always get a user that needs N + 1 new mods. They also might want to use proper names for their mods, e.g. if they're planning to submit them to Unimod, or especially if they just want to use a more up-to-date Unimod file than we include in OpenMS.

**Comment by hendrikweisser** [comment]
> I don't think this would currently work with OpenMS, because the parameter values are validated by the TOPP tool. Wouldn't you get an error about an "illegal parameter value" or something to that extent?

It occurred to me that this should actually not be the case, because after updating the Unimod definitions you'd rebuild OpenMS. So the TOPP tools would know about the extended list of modifications, only KNIME wouldn't.

**Comment by timosachsenberg** [comment]
yes this is primarily a KNIME question

**Comment by hendrikweisser** [comment]
I like Timo's suggestion here: https://github.com/OpenMS/OpenMS/issues/1949#issuecomment-212119599

If you combine that with the option of passing the modification definition file(s) via an advanced parameter, that should also solve the workflow sharing issue (because of course you'd have to include the additional input file).

**Comment by timosachsenberg** [comment]
yep but it seems quite involved and we need to figure out if this is easily possible in KNIME (as we need to change the behavior of the dialog that allows editing of the string list.

**Comment by timosachsenberg** [comment]
A quick workaround would be to add e.g. modifications CUSTOM1,..., CUSTOM10 to our version of unimod so they appear in the StringList. User can then modify it to add their modifications.

**Comment by jpfeuffer** [comment]
After this problem came up again, I reread this issue and also like Timo's suggestion about "dynamic" restrictions. If I find some time I will see what we can do. It might even be possible without an update of our tool description schemata, if we maybe allow a token in string restrictions to mean arbitrary string (at own risk). OTherwise we would need another attribute for restricted parameters.

**Comment by jpfeuffer** [comment]
It is still an important task. But might be solved sufficiently with the dynamic node generation.
I will close once our Plugin ships with this method.

**Comment by hendrikweisser** [comment]
@jpfeuffer:
> I will close once our Plugin ships with this method.

Any update?

**Comment by jpfeuffer** [comment]
Yeah so, the dynamic generation is possible now. I did not switch the deployment of our plugin to using that method yet since it makes the new plugin incompatible with workflows of the old one (you would need to rebuild workflows).
But: After thinking about it, this would not solve the problem of having to recompile all the tools with the modified unimod.xml, which is in my opinion the thing that should be avoided mainly.
It would allow to drag and drop newly compiled executables, yes, but when someone is using KNIME they barely ever have a checked-out source ready to exchange executables on the fly. Actually the same is the case for TOPPAS (if someone uses the installer).
So I think the main problem to fix is in our code and our CTDs. To have the possibility of having some suggested default values that can be listed, but also being able to add user-defined stuff.
Things to do:
- Add flag to registerStringListParam to disable validation
- Add this flag to the internals of the Param object
- Mark it accordingly in the write_ctd option of the tools
- Adapt parameter reading to obey the flag
- Adapt KNIME (and maybe TOPPAS) StringList dialogues to allow a Custom field in the dropdown menu that then adds a text box, where users can type their added modification.

I think the title of the issue is well-written and it can be kept like this. I think implementation effort is low to medium if I am not missing anything.
